### ISSUE TITLE
storage: query hangs when using INSERT FROM ... SELECT on same table

### ISSUE THREAD

**Issue by solongordon** [cockroachdb/cockroach#28842]
If you select from a table and insert the results into the same table, the query never returns, provided the result set is sufficiently large.

To reproduce:
```
CREATE TABLE t (x INT);
INSERT INTO t SELECT generate_series(1, 25000);
INSERT INTO t SELECT * FROM t;
```
The last query never completes. ~~Based on the logs, it seems like this happens when the insert is large enough to trigger a range split, and then it causes an endless cycle of range splits.~~ The splits were a symptom, not the cause. See discussion below.

This issue is present in both 2.0 and 2.1.

**Comment by nvanbenschoten** [comment]
This is interesting. We're intentionally inducing a recursive insertion. Is that valid? Are statements allowed to see values written in the same statement? I wonder what Postgres does.

**Comment by nvanbenschoten** [comment]
I assume this only happens when the `generate_series` is large enough to cause the plan nodes to start chunking retrieval and insertion. Once that starts, later chunks will begin to observe the output of previous chunks and enter an infinite loop. (_still speculation_)

**Comment by nvanbenschoten** [comment]
> I wonder what Postgres does.

This seems to work fine in Postgres.

> Are statements allowed to see values written in the same statement?

@knz could you weigh in here?

**Comment by knz** [comment]
> Are statements allowed to see values written in the same statement?

yes

**Comment by knz** [comment]
(needs to be checked)

**Comment by nvanbenschoten** [comment]
> > Are statements allowed to see values written in the same statement?
>
> yes

If that's the case then this should be an infinite loop. In Postgres this simply doubles the number of rows in the table, which implies that subqueries do not observe values written in the same statement.

**Comment by knz** [comment]
the issue really is to determine the consistency model. A single statement in pg that invokes a stored procedure will allow subsequent statements *inside the stored procedure* to see previous writes.

So the "step" of read-your-writes is not the top-level statement in pg's semantics, but something finer.

**Comment by nvanbenschoten** [comment]
> So the "step" of read-your-writes is not the top-level statement in pg's semantics, but something finer.

Right, but the problem here is that we're not executing the subquery fully before reading our own writes from the top-level statement.

**Comment by knz** [comment]
is that the problem though? do you have evidence that pg loads all the rows of the subquery in a temp table or something similar before performing the insert?

**Comment by petermattis** [comment]
Feels like there needs to be a synchronization point between the reads and the writes in a query like this. If we could buffer the read in a temp table we'd have behavior similar to postgres.

Also, reminds me of [the Halloween Problem](https://en.wikipedia.org/wiki/Halloween_Problem).

**Comment by knz** [comment]
yep looks like a halloween situation :)

For the record, postgres does **not** use a temp table


```
kena=> create table t(x int);
CREATE TABLE
kena=> explain(Verbose) insert into t table t;
                              QUERY PLAN
----------------------------------------------------------------------
 Insert on public.t  (cost=0.00..35.50 rows=2550 width=4)
   ->  Seq Scan on public.t t_1  (cost=0.00..35.50 rows=2550 width=4)
         Output: t_1.x
(3 rows)
```

**Comment by knz** [comment]
I am starting to get a feel of what postgres is doing, thanks to an error I was asked to look at two months ago:

```
kena=> create table u(x int primary key);
CREATE TABLE
kena=> insert into u(x) values (1), (1) on conflict(x) do update set x = excluded.x+1;
ERROR:  ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
```

My working hypothesis is that:

1) a table scan in pg will skip over values inserted within the same statement. 
2) if pg detects that an insert/update is attempting to update a row that was already inserted eearlier *in the same statement* it produces the aforementioned error

The error is meant to guarantee clients do not mis-exploit the apparent "sequence point" behavior introduced by the mechanism in 1.

**Comment by knz** [comment]
(Regarding "skip over values inserted within the same statement" -- we can have a "statement count" within a txn and have the row reader / table reader skip over any rows with the same counter value as the statement performing the scan)

**Comment by bdarnell** [comment]
> we can have a "statement count within a txn" on the intent and have the row reader / table reader skip over any rows with the same counter value as the statement performing the scan

Intents have a sequence counter (counting kv requests, not statements). Could that be part of the solution?

https://github.com/cockroachdb/cockroach/blob/15e4c8dec78aef0bfc8342029a611b4a82cfa4ab/pkg/storage/engine/enginepb/mvcc3.proto#L59-L64

**Comment by knz** [comment]
If the intent is updated per access and not per statement it's hard to determine whether a particular sequence number was generated by the current statement or another before that.

We'd need to collect the current value of that field for every intent laid. We may as well collect just the written PK without the intent which would telll us the same.

**Comment by knz** [comment]
From Ben:

> FWIW the sql standard says that the SELECT query "is effectively evaluated before insertion of any rows into T". so `INSERT INTO t SELECT * FROM t` is well-defined and doubles the size of the table
> although supporting self-referencing operations is optional ("feature F781")
> section 14.11 of the "foundation" pdf in sql 2008
> general rule 5

**Comment by knz** [comment]
The SQL standard and postgres are consistent on that point so we need to get there too.

There are two working directions:

- extend the intent format to include the statement counter inside the txn, and compare that upon KV accesses:
  - during reads, skip over row if counter same as requesting stmt
  - during writes, error out with "duplicate write to same row"

- inside the SQL mutation code keep a set data structure (a hashmap?) of all the PKs that have been touched by the current mutation. Compare it during mutations. This table needs to escape to disk because it can grow large.

**Comment by knz** [comment]
The problem with the 2nd solution (SQL level) is that the SELECT will still see the values being inserted.

This can cause two problems:

- if the SELECT "amplifies" what it reads (e.g. via some windowing)  the mutation-level check will not be sufficient to prevent ballooning RAM usage

- (more serious) if the query applies the SELECT clause to read "old values" with certain assumptions (e.g. that all the values are positive) and then writes "new values" that violate these assumptions (e.g. negative values), the SELECT can start failing, or compute bogus results, when it sees the dirty read (uncommitted write).

Let us not under estimate the severity of this 2nd problem. I am not concerned with cases where the SELECT would fail with an error because some dirty read pops up an  error-generating value (e.g. zero yields division by zero). It's much more dangerous that the dirty read pops up  a value that, after a projection, turns into bogus inserted data **silently**. That's is not client-detectable typically.

**Comment by nvanbenschoten** [comment]
Like Ben mentioned, we already have a sequence count on each request, but this is not exposed to users of `client.Txn`. We could expose this and allow clients to specify the sequence that a request be run at, but this alone still would not be very useful. To make this behave the way we'd want, we'd need to make reads at sequences beneath intents ignore those intents instead of throwing replay errors (which is related to idempotency).

**Comment by knz** [comment]
In order to make a fully working SQL solution, the set of PKs that have been touched by mutations must be used to filter data by all the table readers for that query.

**Comment by knz** [comment]
@nvanbenschoten see my point above -- it's not so clear what number to use when there are multiple statements inside a single txn.

**Comment by nvanbenschoten** [comment]
Which point above? Why can't we just run the read phase of a statement at seq `s` and the write phase of a statement as seq `s+1`? A statement should see the result of all previous statements.

**Comment by nvanbenschoten** [comment]
Of course, there are other concerns with making the sequence number of requests configurable.

**Comment by knz** [comment]
@nvanbenschoten I think (reading ben's comment above) that the field is incremented by 1 on every request.  What you suggest is having the KV client able to overwrite/specify the sequence number. That's the initial solution I proposed and where Peter said "maybe not so late in 2.1".

**Comment by petermattis** [comment]
Is the sequence counter sufficient? We can't just ignore the intent because the sequence number is too new as we might need to read a previously written value in the txn. I think we'd need to keep an intent per sequence point.

Consider:

```
BEGIN
COMMIT
```

**Comment by knz** [comment]
Peter The sequence counter should be equal to the statement counter inside the SQL-level txn. Then it would work

**Comment by petermattis** [comment]
> Peter The sequence counter should be equal to the statement counter inside the SQL-level txn. Then it would work

Right, but that's not how the intent sequence number works right now. We don't keep around the old value.

**Comment by nvanbenschoten** [comment]
> I think we'd need to keep an intent per sequence point.

Yes, absolutely, to make that solution generally usable, we'd need to maintain multiple intents on a single key. This is discussed further in https://github.com/cockroachdb/cockroach/issues/26915.

**Comment by knz** [comment]
Which "old value"? The KV client (SQL) would provide the counter for the current statement. The KV reader would then compare what is provided by SQL and what is found in the intent and decide based on that:

- for reads, skip if they match, accept if the intent counter is lower than the provided value
- for writes, take the provided value from SQL and put it into the intent

**Comment by nvanbenschoten** [comment]
@knz the problem is if there was an intent from a previous statement that has been clobered by a write in the new statement. The "read-phase" of a statement should always see the effects of the "write-phase" of a previous statement.

**Comment by knz** [comment]
that would work with what I'm saying?

statement A has counter value N

statement B has counter value N+1

statement A writes intent with counter value N

statement B reads with "Current stmt counter N+1", because the intent has value N it is "older" and this is accepted.

**Comment by nvanbenschoten** [comment]
@knz want to take this point offline?

**Comment by knz** [comment]
Nathan and I sat together and talked this though.

Here is the outcome, Nathan did most of the correctness checking and helped write the notes.

tldr:
- the KV level solutions discussed above don't work (see below for details)
- a SQL level solution is workable
- general case may come at a slight perf hit, but we can devise fast paths

### KV level solution

- scope of discussion: multiple statements inside a single SQL txn, so all intents are linked to the same txn record
- idea was to extend the idea of "sequence" (the current intent field) to follow the count of stmts inside the sql txn.

Problem with this idea:

- statement A writes data X at counter N
- statement B starts, immediately updates X, bumps sequence number to N+1
- statement B, later, needs to read X "as written by A", but sees an intent with its own counter val and skips over it.

Without storing multiple intents on a single key, this idea is thus not workable.

(Storing multiple intents on a single key can be a separate project, would probably even help with savepoints, but that is larger work and thus not workable for 2.1)

### SQL level solution

tldr: More or less Peter's idea "let's buffer all the things"

- during a mutation, mutated rows are stored in a DiskRowContainer, and do not go to KV
- FK checks also performed during the mutation, use KV data only and not the data in the DiskRowContainer (or maybe not? this needs to be double checked: a user was complaining lately that we should support self-references. pg does it.)
- only when the data source is exhausted, the mutation starts to perform the writes proper, by moving the rows from the DiskRowContainer to KV (and updating the secondary indexes, etc)

### Performance considerations of the SQL solution

Things that tell us this may not be too painful perf wise:

- In the OLTP case with just 1 or very few row modified, this overhead should be negligible: the DiskRowContainer only puts stuff on disk when it grows "large" -- **this needs to be verified**
  - Nathan suggests we need to double check what happens with large VALUES clauses with many rows
- when there are more rows modified, the cost would be 1) extra RAM+disk usage for the temp data 2) another scan of all the mutated rows. However Nathan suggests that the cost of the writes will outweigh the overhead of the 2nd scan. This needs to be checked.
- doing this buffering will incidentally open up new optimization avenues, because this will enable us to re-order the writes and batch them more effectively.

In case we find out we need to care about performance we can do the following:

- devise a predicate that identifies "simple data sources" for a mutation, and disable the buffering in that case
  - Nathan suggests we should prioritize a special case for simple VALUES clauses, especially if we plan to backport this into 2.1

- use the "span collector" machinery that was invented/designed for parallel stmt execution, and use that to enable/disable the buffering (buffering required iff the spans for the data source overlap with the output sink - target table and FK targets)


(A very small ghetto problem with this last point: `CREATE TABLE foo AS SELECT * FROM system.eventlog`, the system.eventlog is a target of the mutation but will not appear in the span collector. Maybe we can ignore this?)

### Prioritization of the solution on the roadmap

Nathan suggests "this fix should be 2.2, what do we do for 2.1"?

knz:
- the KV-level fix with multiple intents per key definitely should go to 2.2 or later
- the SQL solution may actually be small compared to the discussion around the problem, so could qualify for a backport into 2.1
  - if we do this, be mindful of the VALUES clause perf (see comment above)
- it's not like we have a choice really until we have a better "in-between" solution for 2.1

**Comment by andreimatei** [comment]
@knz is it true that, although reads and writes alternate when executing the query in question, there are no reads being done concurrently with writes? As in, when there's a read in progress, the writer is blocked on the respective `tableReader` (which is in turn blocked on a `rowFetcher`)?

**Comment by knz** [comment]
It's currently true yes but will not remain true once we start distributing the mutations.

**Comment by knz** [comment]
Independently from Andrei's question, a followup on yesterday's writeup:

In any case the output of RETURNING must wait until the rows hit KV -- to check both for FK cascading updates/drops and duplicate row checks.

**Comment by andy-kimball** [comment]
I think multiple intents per key is the right long-term architectural solution. I think there would be multiple dividends from allowing multiple intents, like avoiding the halloween problem without buffering, as well as enabling immutability and clean replay. AcidLib did this, and it worked really well. If we ever embark on big architectural changes to Core, this is something to keep in mind.

**Comment by knz** [comment]
Heads up: the KV work to help with this is performed in #33244; this must be complemented by SQL-level work too, the tracking issue for this is #33473.

**Comment by mattcrdb** [comment]
A forum user ran into this limitation recently:

https://forum.cockroachlabs.com/t/performence-of-use-insert-select-statement-to-copy-a-table/3202
